### ISSUE TITLE
Proposal - Remove index.d.ts and consolidate @types/react-native-maps

### ISSUE THREAD

**Issue by methodbox** [react-native-community/react-native-maps#2847]
### Is this a bug report?
No. This is a proposal.

`@types/react-native` contains multiple types that conflict with this module.

There are types in your own index.d.ts that don't match their own definitions in the `react-native` definition; the property `initialRegion`, for example, doesn't exist in the `@types/react-native`, but a definition for `MapView` does, yet doesn't match your own definitions. 

As @types/react-native is practically required to use TS properly with RN, this results in your index.d.ts being overridden and causes a conflict. Rather than update both @types/react-native with definitions that no longer apply to the core RN module *and* match them to your types, consolidation seems appropriate.

My proposal is to remove the `index.d.ts` from this module and consolidate it's definitions into an @types module instead.

The PR @types is to remove `MapView` content entirely as it's not part of the core `react-native` module any longer and conflicts with this (your) module types definitions.

The default recommendation from the `react-native` team is to use @types for TS definitions. Having an `index.d.ts` in your module has the potential to cross name spaces (such as your generic use of the word "Camera") in other type definitions. When moving into the @types module, I would rename this to something like `MapCamera`.

Consolidating this @types would prevent this from happening as, since they are developed under the same repo, it is easy to identify namespace conflicts. Consolidation also has the benefit of maintaining congruence with `react-native` types where your individually maintained type definitions do not.

You can compare the conflicting types here:

https://github.com/methodbox/DefinitelyTyped/blob/master/types/react-native/index.d.ts

The PR I'm working on is here:

https://github.com/DefinitelyTyped/DefinitelyTyped/pull/35239

**Comment by rborn** [comment]
@alvelig @christopherdro any thoughts on this?

**Comment by christopherdro** [comment]
Sounds like a great idea.

**Comment by methodbox** [comment]
I have reviewed the types on this companion module for `react-native-maps` module `...directions`: https://github.com/bramus/react-native-maps-directions/blob/master/index.d.ts and they've merged my fix  there; this in part how I discovered the conflict.

If you guys are good with this, I can submit a PR, but I would hope it wasn't merged until I'm able to complete the PR to create the new `@types/react-native-maps-directions` module.

You should know _that effort_ is pending the approval of another PR with @types:  https://github.com/DefinitelyTyped/DefinitelyTyped/pull/35274 - this due to effectively the same issue, because `@types/styled-components` contains a reference to the deprecated version of `MapView` types here:  https://github.com/methodbox/DefinitelyTyped/commit/78685600044ecee8d8bdc98426829affdf72ff0a

**Comment by methodbox** [comment]
Just an update that a related PR has since been merged (`react-native-maps-directions`), another, which is a dependency on the change to @types/`react-native` (@types/`styled-components`) has been approved and pending merge.

Once that is complete I will reattempt the original PR for removal of these types from @types/`react-native`.

At that point I will create a new @types for `react-native-maps`.  Assuming that is approved, then would probably be the best time for the PR here.

**Comment by methodbox** [comment]
FYI: all my checks just passed, so this will probably be merged into @types soon:

https://github.com/DefinitelyTyped/DefinitelyTyped/pull/35239

I will submit the PR to remove the `index.d.ts` here as soon as I see it merged.

@rborn @christopherdro 

I did not include attributions due to a potential conflict, but this can be updated after it's merged. Apparently, thought, the way the `*.d.ts` header works is this results in notifications on that repo when changes occur.

Let me know if you'd like me to attribute types to specific author(s) and I'll do a PR if needed once it's merged.

**Comment by rborn** [comment]
@methodbox honestly I don't really care about attribution. If that is a potential deal breaker I'd just publish without and then (if needed/people complaining/etc maybe try to fix it)
my 2c.


In other line of thoughts, will the new types package be made a dep of this one or how will people know they need to install it? 

🤗

**Comment by methodbox** [comment]
@rborn It’s merged now, so I can test how it works, which I’ve started doing, but I wouldn’t make changes here quite yet.

I think adding it as a devDep is best or you could add a npm warning but I think those suck and really should be for breaking deps.

**Comment by rborn** [comment]
@methodbox devDep I also think works the best, the easier for the user, the better 😊

**Comment by methodbox** [comment]
> @methodbox devDep I also think works the best, the easier for the user, the better 😊

Ok, well I've tested the new @types definitions which work fine (as they should since they're effectively identical to the ones here) in a number of projects I'm working on and test projects I created for this purpose.

I'm going to put in the PR then to add the `devDependency` and remove `index.d.ts`.

I'm reviewing if we also need to add "resolutions" to operate correctly per this:

https://github.com/DefinitelyTyped/DefinitelyTyped/issues/20350

If anyone has any insight on this, let me know. Testing this atm.

**Comment by methodbox** [comment]
@rborn Sorry, I haven't published an NPM module before, so I'm a bit confused.

If I add `@types/react-native-maps` as a devDep, of course it will install for those working on this module, but it doesn't advise any user that it exists.

I could add it as a peerDependency but it's not explicitly required for the package.

Just want to be sure the route you want to take with this. I'm going to submit the PR with what we discussed already, and I guess I can add the peerDep if that's the way to go.

**Comment by rborn** [comment]
@methodbox nice job 😻 

I'm not versed in this either, but I think that if it's added in`devDependencies` it will be installed with the package so the types will be available to the users. I'm afraid that if we add it to `peerDependencies` on install the users might get a warning but they won't bother to install and this will create a new rain of issues "types don't work"

@christopherdro @kelset can you guys please give us some little advice? 🤗

**Comment by kelset** [comment]
I don't have experience on this, sorry - but looking at react-navigation, it seems that they add it to devDeps -> https://github.com/react-navigation/react-navigation/blob/master/package.json

**Comment by mcmar** [comment]
Will this require every consumer of rn-maps to also `yarn add --dev @types/react-native-maps` manually?

**Comment by mcmar** [comment]
There's a discussion happening on the PR related to this issue: https://github.com/react-native-community/react-native-maps/pull/2877

I think I've explained how there's nothing to be gained from moving the types from here to `@types/react-native-maps`. DefinitelyTyped appears to recommend against this and even recommends deleting the DT module once types are added to the main repo. https://github.com/DefinitelyTyped/DefinitelyTyped#removing-a-package

The OP's original issue was simply related to the expo package and an error in how they re-export types from this library. It has nothing to do with `react-native-maps` and will not be solved by moving the types to `@types/react-native-maps`

**Comment by methodbox** [comment]
> There's a discussion happening on the PR related to this issue: #2877
> 
> I think I've explained how there's nothing to be gained from moving the types from here to `@types/react-native-maps`. DefinitelyTyped appears to recommend against this and even recommends deleting the DT module once types are added to the main repo. https://github.com/DefinitelyTyped/DefinitelyTyped#removing-a-package
> 
> The OP's original issue was simply related to the expo package and an error in how they re-export types from this library. It has nothing to do with `react-native-maps` and will not be solved by moving the types to `@types/react-native-maps`

It will actually be solved by doing this because Expo relies on _this repo_ for their `react-native-maps` (version 0.24 will be in the next Expo release).

However, as you seem unwilling to back down on this, maybe you have a suggestion that also considers the testing that occurs with @types and Travis testing against (I dunno - hundreds?) of `react-native-x` modules?

**Comment by mcmar** [comment]
> It will actually be solved by doing this because Expo relies on this repo for their react-native-maps (version 0.24 will be in the next Expo release). But as I've already explained multiple times, this was tested, not just with Expo.

How expo imports the typedefs for `react-native-maps`, whether thats from `react-native-maps/index.d.ts` or `@types/react-native-maps` doesn't affect the order that they export it in.

> @mcmar seems to keep stating it's just Expo, but I produced the error with plain 'ol React Native, too. Ultimately, this isn't up to me - it's up to the people who maintain this repo.

Dear maintainers: please do not merge the PR. It doesn't fix any issue and only makes life harder for developers who consume this library.

> However, as you seem unwilling to back down on this, maybe you have a suggestion that also considers the testing that occurs with @types and Travis testing against (I dunno - hundreds?) of react-native-x modules?

The DefinitelyTyped repo just uses plain old `dtslint` for testing. We can use that too in this repo and it will do the exact same thing here that it does there. They don't test your modules "against other `react-native-x` modules" because there's no reason to. They're under separate module namespaces.

> I want to clarify my suggestion here wasn't just to solve my problem - it was also to find a better long-term solution. Rather than attack any mistakes I may have made, maybe make a suggestion that fulfills that goal?

I'm not attacking you or your mistakes. There is no solution needed for this repo because there's no problem in this repo. The solution is to open an issue on the `expo` repo.

**Comment by mcmar** [comment]
I tested this on my own with the old `@types/react-native` and `react-native-maps`. This is the exact setup that I've been using in a production app for several months with no conflicts.

I did not see any conflicts and in no case did it pull the types from `@types/react-native` when I do `import { X } from 'react-native-maps';`.

Typescript does not pull typedefs from the `react-native` module namespace when you do `import { X } from 'react-native-maps';`. They're under separate module namespaces. This is documented here: https://www.typescriptlang.org/docs/handbook/namespaces-and-modules.html

**Comment by methodbox** [comment]
> The DefinitelyTyped repo just uses plain old `dtslint` for testing. We can use that too in this repo and it will do the exact same thing here that it does there. They don't test your modules "against other `react-native-x` modules" because there's no reason to. They're under separate module namespaces.

This is incorrect. The Travis tests install a ton of modules (@types definitions for a slew of RN modules) and check for conflicts.

As I linked to in the PR, here is an example from my merged PR to remove the deprecated types: https://travis-ci.org/DefinitelyTyped/DefinitelyTyped/builds/534054337?utm_source=github_status&utm_medium=notification

I also want to restate some misinformation you've posted multiple times here: `react-native-maps` was separated into it's own module but `MapView` was deprectated long ago and removed from RN quite a while ago, too.

There's no reason those definitions needed to be in there. However, I think it's worth restating here (as in the PR - this is silly) that _had the types for this module been in @types before_ any potential conflicts (like those in Expo, if we want to use that example we both agree on) would have been found.

As to what the solution would have been in that case, we can certainly debate but knowing the _cause_ certainly would have been valuable.

**Comment by rborn** [comment]
Closing in favor of PR https://github.com/react-native-community/react-native-maps/pull/2877
